eth-riscv-runtime/src/lib.rs

```diff
@@ -116,6 +118,55 @@ pub fn msg_sender() -> Address {
     Address::from_slice(&bytes)
 }
 
+pub fn msg_value() -> U256 {
+    let first: u64;
+    let second: u64;
+    let third: u64;
+    let fourth: u64;
+    unsafe {
+        asm!("ecall", lateout("a0") first, lateout("a1") second, lateout("a2") third, lateout("a3") fourth, in("t0") u64::from(Syscall::CALLVALUE));
+    }
+    U256::from_limbs([first, second, third, fourth])
+}
+
+pub fn msg_sig() -> [u8; 4] {
+    let first: u64;
+    let offset: u64 = 0;
+    unsafe {
+        asm!("ecall", lateout("a0") first, in("a0") offset, in("t0") u64::from(Syscall::CALLDATALOAD));
```

Comment:
So regarding calldata, currently we dont implement it as a host function, but rather inject it directly as a ROM for the interpreter which lets the contract read "for free". We need to decide which approach we want to use, but I quite like this one.

Reply:
Reading from ROM should be faster than `syscalls`.  Let's change it back to that?

Reply:
Yep sounds good